internal/app/machined/pkg/controllers/secrets/etcd_test.go

```diff
@@ -0,0 +1,132 @@
+
+	etcdCA, err := x509.NewSelfSignedCertificateAuthority(
+		x509.Organization("talos"),
+		x509.ECDSA(true),
+	)
+	suite.Require().NoError(err)
+
+	rootSecrets.TypedSpec().EtcdCA = &x509.PEMEncodedCertificateAndKey{
+		Crt: etcdCA.CrtPEM,
+		Key: etcdCA.KeyPEM,
+	}
+	suite.Require().NoError(suite.State().Create(suite.Ctx(), rootSecrets))
+
+	networkStatus := network.NewStatus(network.NamespaceName, network.StatusID)
+	networkStatus.TypedSpec().AddressReady = true
+	networkStatus.TypedSpec().HostnameReady = true
+	suite.Require().NoError(suite.State().Create(suite.Ctx(), networkStatus))
+
+	hostnameStatus := network.NewHostnameStatus(network.NamespaceName, network.HostnameID)
+	hostnameStatus.TypedSpec().Hostname = "host"
+	hostnameStatus.TypedSpec().Domainname = "domain"
+	suite.Require().NoError(suite.State().Create(suite.Ctx(), hostnameStatus))
+
+	nodeAddresses := network.NewNodeAddress(network.NamespaceName, network.FilteredNodeAddressID(network.NodeAddressRoutedID, k8s.NodeAddressFilterNoK8s))
+	nodeAddresses.TypedSpec().Addresses = []netaddr.IPPrefix{
+		netaddr.MustParseIPPrefix("10.3.4.5/24"),
+		netaddr.MustParseIPPrefix("2001:db8::1eaf/64"),
+	}
+	suite.Require().NoError(suite.State().Create(suite.Ctx(), nodeAddresses))
+
+	timeSync := timeres.NewStatus()
+	timeSync.TypedSpec().Synced = true
+	suite.Require().NoError(suite.State().Create(suite.Ctx(), timeSync))
+
+	suite.AssertWithin(10*time.Second, 100*time.Millisecond, func() error {
+		certs, err := ctest.Get[*secrets.Etcd](
+			suite,
+			resource.NewMetadata(
+				secrets.NamespaceName,
+				secrets.EtcdType,
+				secrets.EtcdID,
+				resource.VersionUndefined,
+			),
+		)
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				return retry.ExpectedError(err)
+			}
+
+			return err
+		}
+
+		etcdCerts := certs.TypedSpec()
+
+		serverCert, err := etcdCerts.Etcd.GetCert()
+		suite.Require().NoError(err)
+
+		suite.Assert().Equal([]string{"host", "host.domain", "localhost"}, serverCert.DNSNames)
+		suite.Assert().Equal("[10.3.4.5 2001:db8::1eaf 127.0.0.1 ::1]", fmt.Sprintf("%v", serverCert.IPAddresses))
+
+		suite.Assert().Equal("host", serverCert.Subject.CommonName)
+
+		peerCert, err := etcdCerts.EtcdPeer.GetCert()
+		suite.Require().NoError(err)
+
+		suite.Assert().Equal([]string{"host", "host.domain"}, peerCert.DNSNames)
+		suite.Assert().Equal("[10.3.4.5 2001:db8::1eaf]", fmt.Sprintf("%v", peerCert.IPAddresses))
+
+		suite.Assert().Equal("host", peerCert.Subject.CommonName)
+
+		adminCert, err := etcdCerts.EtcdAdmin.GetCert()
+		suite.Require().NoError(err)
+
+		suite.Assert().Empty(adminCert.DNSNames)
+		suite.Assert().Empty(adminCert.IPAddresses)
+
+		suite.Assert().Equal("talos", adminCert.Subject.CommonName)
+
+		kubeAPICert, err := etcdCerts.EtcdAPIServer.GetCert()
+		suite.Require().NoError(err)
+
+		suite.Assert().Empty(kubeAPICert.DNSNames)
+		suite.Assert().Empty(kubeAPICert.IPAddresses)
+
+		suite.Assert().Equal("kube-apiserver", kubeAPICert.Subject.CommonName)
```

Comment:
shouldn't we add a test where we update the node addresses and see the certs sans got updated?

Reply:
yep, I can do that... should be fairly easy

Reply:
done